Update origin documentation, including first pass at S3 config (#1086)
PelicanPlatform/pelican: docs/pages/serving_an_origin.mdx

Reviewer: I feel like the section covering "Launch the Origin With an S3 Storage Backend" should be moved to a separate page, as currently it blocks the flow from running an origin to setting up admin website and checking if everything is working, and make the documentation a bit long to follow

Author: Do our docs support any accordion drop downs? Maybe we can keep the section collapsed by default?

For _Linux_ users, it is recommended to install Pelican using one of the package managers (RPM, APK, Deb, etc.) so that Pelican dependecies are automatically handled. You may also run a [Pelican docker image](./install/docker.mdx) to serve a Pelican origin.
If you prefer to install Pelican as a standalone binary, you need to follow [additional instructions](https://osg-htc.org/docs/data/xrootd/install-standalone/#install-xrootd-standalone) to install dependencies for the Pelican origin server.

> **NOTE:** Serving origins with a standalone Pelican binary is possible, but not recommended.

For _macOS_ and _Windows_ users who want to serve a Pelican origin, please use [Pelican docker image](./install/docker.mdx).

### Open Firewall Port for Pelican Origin

> At their core, Pelican origins are web servers that listen to two TCP ports for file transfers and Web UI. By default, the browser/API interface for your origin will be at port `8444`, and the port for object transfers will be at `8443`. You may change these port numbers through the [configuration file](./parameters.mdx) with parameters [`Origin.Port`](./parameters.mdx#Origin-Port) and [`Server.WebPort`](./parameters.mdx#Server-WebPort) respectively.

Reviewer: ```suggestion
At their core, Pelican origins are web servers that listen to two TCP ports for file transfers and Web UI. By default, the Web UI and API for your origin will be at port `8444`, and the port for object transfers will be at `8443`. You may change these port numbers through the [configuration file](./parameters.mdx) with parameters [`Origin.Port`](./parameters.mdx#Origin-Port) and [`Server.WebPort`](./parameters.mdx#Server-WebPort) respectively.
```

In order for Pelican origins to work properly, these ports need to be accessible by the federation, which in most cases means they need to be open to the internet. If your server host has a firewall policy in place, please open these two ports for both incoming the outgoing TCP requests.

> > **NOTE:** If it is not possible for you to expose any ports through the firewall (eg you're on a local network or behind a NAT), Pelican has a special feature called a _Connection Broker_ that allows you to serve origins without publicly-accessible ports and TLS credentials. However, this is an experimental feature and requires the Pelican federation you are joining to be compatible. If you are interested in learning more about the Connection Broker, please contact help@pelicanplatform.org for further instructions.

Reviewer: ```suggestion
> **NOTE:** If it is not possible for you to expose any ports through the firewall (e.g. you're on a local network or behind a NAT), Pelican has a special feature called _Connection Broker_ that allows you to serve origins without publicly-accessible ports and TLS credentials. However, this is an experimental feature and requires the Pelican federation you are joining to be compatible. If you are interested in learning more about the Connection Broker, please contact help@pelicanplatform.org for further instructions.
```

Since your TLS certificate is associated with your domain name, you will need to change the default hostname of Pelican server to be consistent. Set `Server.Hostname` to your domain name (e.g. `example.com`).

> ## Serving Origins

Reviewer: ```suggestion
## Serve Origins
```

I thought we want to keep titles/subtitles started with a verb ( from Brian Lin's instructions on written a technical docs)

Author: I think this still aligns with the style formatting Brian prefers. Take a look at the current OSDF documentation he and Matyas have written -- there they also begin headers with gerunds like "Running", "Configuring", "Providing" etc.

When you've completed the aforementioned steps, you're ready to start configuring the origin that will add your data to a federation. Serving an origin is the process of taking some underlying storage repository and making its data accessible via a namespace prefix in your federation. For example, you might make files in the directory `/my/directory` available at the federation path `/my/namespace` so that anyone with access to the federation can get objects from the directory

> By default, Pelican origins serve files from a POSIX backend, the filesystem used by Linux computers. However, Pelican aims to support a variety of backends and we currently also support serving objects from S3. Configuration for S3 is mostly similar to configuration for POSIX filesystems, but with a few important differences. For information about S3 backends, jump to our [s3 documentation](#launch-the-origin-with-an-s3-storage-backend)

Reviewer: ```suggestion
By default, Pelican origins serve files from a POSIX backend, the filesystem used by Linux computers. However, Pelican aims to support a variety of backends and we currently also support serving objects from S3. Configuration for S3 is mostly similar to configuration for POSIX filesystems, but with a few important differences. For information about S3 backends, jump to the [S3 documentation](#launch-the-origin-with-an-s3-storage-backend)
```

> If you are running Pelican docker image to serve an origin, please refer to the [docker image documentation](./install/docker.mdx#run-pelican-origin-server).

### Find a federation to join

Before serving an origin, you need to decide which [**federation**](./client-usage.mdx#useful-terminology) your data will be accessed through. For example, the Open Science Data Federation (OSDF) is Pelican's flagship federation, and if you are interesting in serving an OSDF origin, you can refer to the [OSDF website](https://osg-htc.org/) for details about how to join.

Federations are identified by a URL, which is used to host information your origin will need in order to discover other federation services. For example, the OSDF's federation URL is `https://osg-htc.org`, and an origin that joins the OSDF will visit `https://osg-htc.org/.well-known/pelican-configuration` to get important metadata about the federation's central services (the Director and Registry).

To point your origin at a specific federation, you can either pass the `-f <federation URL>` flag if running from the command line, or configure `Federation.DiscoveryUrl: <federation URL>` in your config yaml.

> ### Serve the Origin

Reviewer: I avoid using "serve" here to distinguish between the secondary title for this section "Serve Origins". Up to you to decide which one to use

> Origins can be configured via the command line, via a config file named `pelican.yaml`, via environment variables, or through a combinations of the three. While simple origins can be run entirely from command line arguments, more complex origins will require configuration your your `pelican.yaml`.

Reviewer 2: ```suggestion
Origins can be configured via the command line, a config file named `pelican.yaml`, environment variables, or through a combinations of the three. While simple origins can be run entirely from command line arguments, more complex origins will require configuration your your `pelican.yaml`.
```

To start a simple pelican origin from the command line that serves POSIX data, run:

```bash
pelican origin serve -f <federation URL> -v </path/to/data>:</your/federation/prefix>
```

Where:

* `<federation URL>` is the federation URL discussed above
* `</path/to/data>` is the absolute path to the directory containing files you want to export as Pelican objects in your federation
* `</your/federation/prefix>` is the federation prefix at which files in `/path/to/data` will be accessed from in the federation. Note that federation prefixes follow POSIX path conventions, and they must begin with `/` to denote an absolute path.

> > **NOTE:** By default, origins require authorization tokens for object access. There's currently no way to serve a public origin using only the command line, but you can find more information about various access controls by looking at [origin capabilities](#origin-and-namespace-capabilities) below.

Reviewer: ```suggestion
> **NOTE:** By default, origins require authorization tokens for object access. Pelican currently does not support serving a public origin using the command line, but various access controls can be configured through the configuration file. For more information, see [origin capabilities](#origin-and-namespace-capabilities) below.
```

> To run the same origin using a `pelican.yaml` configuration file, save your configuration to `/etc/pelican/pelican.yaml` if you're running Pelican as root, and at `~/.config/pelican/pelican.yaml` if you're running as a non-root user. The command line origin from above could be configured accordingly:

Reviewer: ```suggestion
To run the same origin using a `pelican.yaml` configuration file, save your configuration to `/etc/pelican/pelican.yaml` if you're running Pelican as root, or at `~/.config/pelican/pelican.yaml` if you're running as a non-root user. The command line origin from above could be configured accordingly:
```

> ```

Reviewer: ```suggestion
```yaml filename="pelican.yaml"
```

# Tell Pelican which federation you're joining
Federation:
  DiscoveryUrl: <federation URL>

# Configure your Origin
Origin:
  # POSIX is the default storage type for Pelican origins
  # and can be omitted
  StorageType: "posix"

  Exports:
    - StoragePrefix: "/path/to/data"
      FederationPrefix: "/your/federation/prefix"
      # Explicitly state what capabilities you want this prefix to have
      Capabilities: ["Reads", "Writes"]

```
and then simply run
```bash
pelican origin serve
```
Pelican will read the config file and apply it to your origin.

> Finally, origins can be configured to a limited extent with environment variables. In Pelican's environment variable model, configuration options are taken from `pelican.yaml`, flattened, and prepended with either `PELICAN_` or `OSDF_`, depending on the name of the binary you're using (ie whether you run `osdf serve` or `pelican serve` commands).

Reviewer: ```suggestion
Finally, origins can be configured to a limited extent with environment variables. In Pelican's environment variable model, configuration options are taken from `pelican.yaml`, flattened, and prepended with either `PELICAN_` or `OSDF_`, depending on the name of the binary you're using (i.e. whether you run `osdf serve` or `pelican serve` commands).
```

For example, you might configure the origin's storage type by setting the environment variable `PELICAN_ORIGIN_STORAGETYPE=posix`.

> Environment variable configuration does not support the same complex structures that can be built with yaml configuration, such as lists.

The first time the origin is started, you will see something that looks like the following:

```console
$ pelican origin serve -f https://osg-htc.org -v $PWD:/demo

Pelican admin interface is not initialized
To initialize, login at https://localhost:8444/view/initialization/code/ with the following code:
551220
```

> See the [browser configuration](#login-to-admin-website) documentation section for more information about initializing your origin's web interface.

Reviewer: ```suggestion
See the [admin website configuration](#login-to-admin-website) documentation section for more information about initializing your origin's admin website.
```

> ### Origin and Namespace Capabilities

Reviewer: Would it worth moving this section and Multi-Export origin to a separate page like "Configuring Origin"? to make the documentation easier to follow?

Capabilities are the configuration options you can assign to origins and namespace prefixes to determine what kinds of access controls you want them to respect. In the previous yaml configuration, we configured the origin capabilities using the `Capabilities` list of the `Origin.Exports` block by specifying that the federation prefix `/your/federation/prefix` supports "Reads" and "Writes". This list of capabilities can be used for further control of what types of operations the namespace is willing to support. Available capabilities include:

- "Reads": When included, objects from the namespace can be read with a valid authorization token.
- "PublicReads": When set, objects from the namespace become public and require no authorization to read.
- "Writes": When included, objects can be written back to the storage backend by Pelican. Write operations _always_ require a valid authorization token.

> - "DirectReads": When included, a namespace indicates that data being pulled from it should only be pulled through a cache and not directly by clients. This may be useful in cases where the origin isn't very performant or has to pay egress costs when data moves through it. Note that this is respected by federation central services, but may not be respected by all clients.

Reviewer: ```suggestion
- "DirectReads": When included, a namespace indicates that data being pulled from it should only be pulled through a cache and not directly by clients. This may be useful in cases where the origin isn't very performant or has to pay egress costs when data moves through it. Note that this is respected by federation central services, but may not be respected by all clients.
```

I thought this is the other way around. If `DirectReads` is not included (by default), shouldn't the director require a cache to be present or fail to redirect?

Author: Hah, good catch.

- "Listings": When included, the namespace indicates it will allow object discovery. Be careful when setting this for authorized namespaces, as this will allow anyone to discover the names of objects exported by this namespace.

> **NOTE:** Most origins should have either "Reads" or "PublicReads" enabled. If neither is set, the origin won't export any data.

There is an important distinction between _origin_ capabilities and _namespace_ capabilities. While it's sometimes easy to treat origins and namespaces as the same thing, Pelican must distinguish between them because two separate origins may export portions of the same namespace, and a single origin may export two disparate prefixes. The only exception to this rule is when a single origin serves a single namespace.

> To configure _origin_ capabilities, you can set top-level options for the origin:

Reviewer: The description in `parameter.yaml` seems to contradict what was stated here: 

> NOTE: This config option is incompatible with multiple exports defined via `Origin.Exports` and requires that the origin exports
  only a single path.

- `Origin.EnableReads`: When true, the origin supports reads that are accompanied by a valid authorization token.
- `Origin.EnablePublicReads`: When true, the origin supports reads by anyone without an authorization token.
- `Origin.EnableWrites`: When true, objects can be written back to the storage backend through the origin. Writes always require a valid authorization token.
- `Origin.EnableDirectReads`: When true, the origin indicates it's willing to serve clients directly, potentially without caching data. Note that this is respected by federation central services, but may not be respected by all clients.
- `Origin.EnableListings`: When true, the origin will allow object discovery.

If no `Origin.Exports` block is provided to Pelican, these values will also be applied to your federation prefix.

> **NOTE:** Pelican tries to resolve differences between origin and namespace configurations by respecting the more restrictive of the two. If you serve an origin that enables public reads, but the underlying prefix it exports disables all reads, you won't be able to read from that namespace.

### Multi-Export Origins

> The previous examples have shown how one might export a single namespace, but Pelican origins can export multiple paths from the same storage backend under different namespaces. For example, you have have two posix directories called `/my/data/public` and `/my/data/private`. If you want to make your public data available under the namespace `/my/prefix/public` and your private data available under `/my/prefix/private`, you'll need to configure a multi-export origin, which is accomplished through the origin's `Exports` block. Below is an example of what that looks like, along with how you could configure access control for the two namespaces:

Reviewer: ```suggestion
The previous examples have shown how one might export a single namespace, but Pelican origins can export multiple paths from the same storage backend under different namespaces. For example, assume you have have two posix directories called `/my/data/public` and `/my/data/private`. If you want to make your public data available under the namespace `/my/prefix/public` and your private data available under `/my/prefix/private`, you'll need to configure a multi-export origin, which is accomplished through the origin's `Exports` block. Below is an example of what that looks like, along with how you could configure access control for the two namespaces:
```

### Serving an S3 Origin

Serving S3 origins with Pelican is similar to serving POSIX origins, but with several key differences. The first is that Pelican must be configured to host an S3 backend, using the configuration option `Origin.StorageType = s3`. To make your work with S3, it needs to know at least four additional things:

- The URL you use to access objects from S3, also known as the _S3 Service URL_
- The _region_ that your S3 instance is hosted out of (almost always `us-east-1` unless you're actually using S3 from Amazon)

> - The name of the bucket your objects are stored in

Reviewer: ```suggestion
- The name of the _bucket_ your objects are stored in
```

```
Origin:
  # Things that configure the origin itself
  # Tell the origin it will be serving objects from S3
  StorageType: "s3"
  S3ServiceUrl: "https://s3.us-east-1.amazonaws.com"
  S3Region: "us-east-1"
  S3UrlStyle: "virtual"

  # The actual namespaces we export. Each export is defined
  # via its own export block
  Exports:
    - FederationPrefix: /aws-public
      Capabilities: ["PublicReads", "Listings", "DirectReads"]
```

In this configuration, users who wish to fetch objects from the origin will still need to know the name of the bucket that hosts those objects. For example, the AWS public bucket `noaa-wod-pds` has an object called `MD5SUMS`, and with this configuration the object can be fetched at `/aws-public/noaa-wod-pds/MD5SUMS`.

## Login to Admin Website

> After your origin is running, the next step is to initialize its web interface, which can be used by administrators for monitoring and further configuration. To initialize this interface, go to the URL specified in the terminal. By default, it should point to https://localhost:8444/view/initialization/code/

Reviewer: ```suggestion
After your origin is running, the next step is to initialize its admin website, which can be used by administrators for monitoring and further configuration. To initialize this interface, go to the URL specified in the terminal. By default, it should point to https://localhost:8444/view/initialization/code/
```
